File: tokenize.js
const wasmCode = Deno.readFileSync(Deno.args[0] ?? "tokenize.wasm");
const wasmModule = new WebAssembly.Module(wasmCode);

let memory = undefined;

let instance;
try {
  instance = new WebAssembly.Instance(wasmModule, {
    env: {},
  });
} catch (error) {
  console.error(error);
}

try {
  const source = Deno.readFileSync("big.zig");
  const ptr = instance.exports.alloc(source.length);
  const bytes = new Uint8Array(instance.exports.memory.buffer);
  const count = instance.exports.tokenize(ptr, source.length);
  instance.exports.free(ptr, source.length);
  console.log(count, "tokens");
} catch (error) {
  console.error(error);
}

File: tokenize.zig
const std = @import("std");
const allocator = std.heap.wasm_allocator;

export fn alloc(len: usize) [*c]u8 {
    return allocator.allocSentinel(u8, len, 0) catch std.debug.panic("oom", .{});
}

export fn free(ptr: [*c]u8, len: usize) void {
    allocator.free(@as([:0]u8, @ptrCast(ptr[0..len])));
}

pub export fn tokenize(source: [*c]u8, len: usize) usize {
    var count: usize = 0;
    var tokenizer = std.zig.Tokenizer.init(@ptrCast(source[0..len]));
    while (tokenizer.next().tag != .eof) {
        count += 1;
    }
    return count;
}

pub fn main() void {}
